src/jit/codegenxarch.cpp

```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
-            // pointer to (fieldOffset + sizeof(fieldType)) to account for the `push`.
-            const int fieldSize     = genTypeSize(fieldType);
-            const int desiredOffset = current->gtFieldOffset + (usePush ? fieldSize : 0);
-            if (currentOffset > desiredOffset)
-            {
-                assert(!preAdjustedStack);
-
-                // The GC encoder requires that the stack remain 4-byte aligned at all times. Round the adjustment up
-                // to the next multiple of 4. If we are going to generate a `push` instruction, the adjustment must
-                // not require rounding.
-                const int adjustment = roundUp(currentOffset - desiredOffset, 4);
-                assert(!usePush || (adjustment == (currentOffset - desiredOffset)));
-                inst_RV_IV(INS_sub, REG_SPBASE, adjustment, EA_PTRSIZE);
-                currentOffset -= adjustment;
-                genStackLevel += adjustment;
-            }
-
-            // Note that the argReg may not be the lcl->gtRegNum, if it has been copied
-            // or reloaded to a different register.
-            const regNumber argReg = fieldNode->gtRegNum;
-            if (usePush)
-            {
-                // Adjust the stack if necessary and push the field.
-                // Push the field.
-                inst_RV(INS_push, argReg, fieldType, emitTypeSize(fieldType));
-                currentOffset -= fieldSize;
-                genStackLevel += fieldSize;
+            // Consume the register, if any, for this field. Note that genConsumeRegs() will appropriately
+            // update the liveness info for a lclVar that has been marked RegOptional, which hasn't been
+            // assigned a register, and which is therefore contained.
+            // Unlike genConsumeReg(), it handles the case where no registers are being consumed.
+            genConsumeRegs(fieldNode);
+            regNumber argReg = fieldNode->gtRegNum;
+
+            // If we can't push this field, it needs to be in a register so that we can store
+            // it to the stack location.
+            if (!m_pushStkArg)
+            {
+                bool needByteReg =
+                    (varTypeIsByte(fieldType) && ((argReg == REG_NA) || ((genRegMask(argReg) & RBM_BYTE_REGS) == 0)));
+                if ((argReg == REG_NA) || needByteReg)
+                {
+                    assert(varTypeIsIntegralOrI(fieldNode));
+                    if (tmpReg == REG_NA)
+                    {
+                        assert(putArgStk->gtRsvdRegs != RBM_NONE);
+                        assert(genCountBits(putArgStk->gtRsvdRegs) == 1);
+                        tmpReg = genRegNumFromMask(putArgStk->gtRsvdRegs);
+                        assert(genIsValidIntReg(tmpReg));
+                    }
+                    if (needByteReg)
+                    {
+                        noway_assert((genRegMask(tmpReg) & RBM_BYTE_REGS) != 0);
```

Comment:
I don't think that this should be a `noway_assert`: the other reg-related asserts above are simple `assert`s, so I don't think this adds much value. Furthermore, I maintain that we should not use `noway_assert`s unless we can prove that they are in fact recoverable.